### PR TITLE
S2U-7 Tests & Quizzes: Assessment Report

### PR DESCRIPTION
**Pull request by stetsche**
Adds:
- New "Skewness coefficient" row to (assessment) Statistics and export to PDF and XLSX
- New "Difficulty", "Total attempts", "Total correct" and "Total incorrect" to item analysis and export to PDF and XLSX

https://sakaiproject.atlassian.net/browse/S2U-7

### CODE REVIEW

File: samigo/samigo-app/src/webapp/jsf/evaluation/detailedStatistics.jsp [hunk 1]
```diff
@@ -95,11 +96,30 @@ $Id: histogramScores.jsp 38982 2007-12-06 13:05:38Z gopal.ramasammycook@gmail.co
 	   <f:valueChangeListener type="org.sakaiproject.tool.assessment.ui.listener.evaluation.HistogramListener" />
      </h:selectOneMenu>
     </h:panelGroup>
+    </div>
+    <div class="dropdown">
+        <button class="btn btn-default dropdown-toggle" type="button" data-toggle="dropdown">
```
[the remark this answers is not in the document]
Comment (stetsche) [reply samigo/samigo-app/src/webapp/jsf/evaluation/detailedStatistics.jsp:101]: false alarm

File: samigo/samigo-app/src/webapp/jsf/evaluation/histogramScores.jsp [hunk 2]
```diff
@@ -99,8 +101,25 @@ $Id$
               <f:valueChangeListener type="org.sakaiproject.tool.assessment.ui.listener.evaluation.HistogramListener" />
             </h:selectOneMenu>
           </h:panelGroup>
-
-          <br /><br />
+          <div class="dropdown">
+            <button class="btn btn-default dropdown-toggle" type="button" data-toggle="dropdown">
```
[the remark this answers is not in the document]
Comment (stetsche) [reply samigo/samigo-app/src/webapp/jsf/evaluation/histogramScores.jsp:105]: false alarm

**Comment by mpellicer** [comment]
Can you paste some screenshots of the feature please? and the reports? Thanks!

**Comment by mpellicer** [comment]
I broke it @stetsche sorry, could you rebase ?

**Comment by stetsche** [comment]
Sure, I fixed it. I had to do a few changes due to conflicts with SAK-48147, but I tested it and it should be fine.

**Comment by stetsche** [comment]
The feature looks very similar to the prototype, but I'll add some screenshots to the Jira tomorrow!

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/EvaluationMessages_ca.properties [hunk 3]
```diff
@@ -381,3 +381,15 @@ answers_title=Respostes
...
+
+# S2U-7
+export_as_pdf=Exporta com a PDF
+export_as_xlsx=Exporta com a Excel
+export_pdf=com a PDF
+export_xlsx=com a Excel
+export=Exporta
+skew_coef=Coeficient d\u0027Asimetria
+difficulty=Dificultat
+total_attempts=Veus Contestada
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/EvaluationMessages_ca.properties:393]:
```suggestion
total_attempts=Vegades Contestada
```

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/EvaluationMessages_ca.properties [hunk 4]
```diff
@@ -381,3 +381,15 @@ answers_title=Respostes
 
 # S2U-9
 print_report=Imprimir reporti
+
+# S2U-7
+export_as_pdf=Exporta com a PDF
+export_as_xlsx=Exporta com a Excel
+export_pdf=com a PDF
+export_xlsx=com a Excel
+export=Exporta
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/EvaluationMessages_ca.properties:390]:
```suggestion
export=Exportar
```

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/EvaluationMessages_ca.properties [hunk 5]
```diff
@@ -381,3 +381,15 @@ answers_title=Respostes
 
 # S2U-9
 print_report=Imprimir reporti
+
+# S2U-7
+export_as_pdf=Exporta com a PDF
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/EvaluationMessages_ca.properties:386]:
```suggestion
export_as_pdf=Exportar com a PDF
```

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/EvaluationMessages_ca.properties [hunk 6]
```diff
@@ -381,3 +381,15 @@ answers_title=Respostes
...
+export_as_xlsx=Exporta com a Excel
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/EvaluationMessages_ca.properties:387]:
```suggestion
export_as_xlsx=Exportar com a Excel
```

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/EvaluationMessages_ca.properties [hunk 7]
```diff
@@ -381,3 +381,15 @@ answers_title=Respostes
...
+export_pdf=com a PDF
+export_xlsx=com a Excel
+export=Exporta
+skew_coef=Coeficient d\u0027Asimetria
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/EvaluationMessages_ca.properties:391]:
```suggestion
skew_coef=Coeficient d\u2019Asimetria
```

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/EvaluationMessages_es.properties [hunk 8]
```diff
@@ -381,3 +381,15 @@ answers_title=Respuestas
...
 # S2U-9
 print_report=Imprimir reporte
+
+# S2U-7
+export_as_pdf=Exportar como PDF
+export_as_xlsx=Exportar como Excel
+export_pdf=como PDF
+export_xlsx=como Excel
+export=Exportar
+skew_coef=Coeficiente de Asimetr\u00eda
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/EvaluationMessages_es.properties:391]:
```suggestion
skew_coef=Coeficiente de asimetr\u00eda
```

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/EvaluationMessages_eu.properties [hunk 9]
```diff
@@ -355,3 +355,15 @@ answers_title=Erantzunak
 
 # S2U-9
 print_report=Inprimatu txostena
+
+# S2U-7
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/EvaluationMessages_eu.properties:359]: @pablosanroman would you mind reviewing these strings please? Thank you very much

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/HistogramListener.java [hunk 10]
```diff
@@ -2282,13 +2302,16 @@ private Map getAssessmentStatisticsMap(List scoreList)
...
-    statMap.put("median", castingNum(calMedian(scores),2));
+	double median = calMedian(scores);
+    statMap.put("median", castingNum(median, 2));
     statMap.put("mode", castingNumForMode(calMode(scores)));
 
     statMap.put("numStudentCollection", numStudents);
     statMap.put(
       "rangeCollection", calRange(numStudents, min, max, interval));
-    statMap.put("standDev", castingNum(calStandDev(scores, mean),2));
+	double standardDeviation = calStandDev(scores, mean);
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/HistogramListener.java:2312]:
```suggestion
    double standardDeviation = calStandDev(scores, mean);
```

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/model/AssessmentReport.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.Optional;
+
+import lombok.Builder;
+import lombok.Data;
+import lombok.Singular;
+
+@Data
+@Builder(builderMethodName = "privateBuilder")
+public class AssessmentReport {
+
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/model/AssessmentReport.java:13]: Remove extra line

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/model/AssessmentReport.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+    private Optional<String> title;
+    private Optional<String> subject;
+    private AssessmentReportType type;
+    @Singular
+    private List<AssessmentReportSection> sections;
+    @Builder.Default
+    private AssessmentReportOrientation orientation = AssessmentReportOrientation.PORTRAIT;
+
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/model/AssessmentReport.java:22]: Remove extra line

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/model/AssessmentReport.java [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    // Builder with mandatory field "type"
+    public static AssessmentReportBuilder type(AssessmentReportType type) {
+        return privateBuilder().type(type);
+    }
+
+    // Make the default builder method private
+    private static AssessmentReportBuilder privateBuilder() {
+        return new AssessmentReportBuilder();
+    }
+
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/model/AssessmentReport.java:33]: Remove extra line

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/model/AssessmentReport.java [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+    public static class AssessmentReportBuilder {
+
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/model/AssessmentReport.java:36]: Remove extra line

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/model/AssessmentReport.java [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+        private Optional<String> title = Optional.empty();
+        private Optional<String> subject = Optional.empty();
+
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/model/AssessmentReport.java:40]: Remove extra line

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/model/AssessmentReport.java [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+            this.title = Optional.ofNullable(title);
+            return this;
+        }
+
+        public AssessmentReportBuilder subject(String subject) {
+            this.subject = Optional.ofNullable(subject);
+            return this;
+        }
+    }
+
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/model/AssessmentReport.java:52]: Remove extra line

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/model/AssessmentReportSection.java [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.Optional;
+
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+public class AssessmentReportSection {
+
+
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/model/AssessmentReportSection.java:14]: Remove extra line

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/model/AssessmentReportSection.java [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+    private Optional<String> title;
+    private List<String> tableHeader;
+    private List<List<String>> tableData;
+    @Builder.Default
+    private TableLayout tableLayout = TableLayout.HORIZONTAL;
+
+
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/model/AssessmentReportSection.java:21]: Remove extra line

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/servlet/evaluation/ExportReportServlet.java [hunk 19]
```diff
@@ -0,0 +1,353 @@
...
+
+    public static final String PARAM_SITE_ID = "siteId";
+    public static final String PARAM_ASSESSMENT_ID = "assessmentId";
+    public static final String PARAM_FORMAT = "format";
+    public static final String PARAM_TYPE = "type";
+    public static final String EXPORT_TYPE_ITEM_ANALYSIS = "item_analysis";
+    public static final String EXPORT_TYPE_STATISTICS = "statistics";
+
+    private static final ResourceLoader EVENT_LOG_BUNGLE = new ResourceLoader(SamigoConstants.EVAL_BUNDLE);
+
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/servlet/evaluation/ExportReportServlet.java:56]: Remove extra line

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/servlet/evaluation/ExportReportServlet.java [hunk 20]
```diff
@@ -0,0 +1,353 @@
...
+import org.springframework.http.ContentDisposition;
+import org.springframework.http.HttpHeaders;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class ExportReportServlet extends SamigoBaseServlet {
+
+
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/servlet/evaluation/ExportReportServlet.java:47]: Remove extra line

**Comment by stetsche** [comment]
@mpellicer Do you mind those extra lines a lot? I got in the habit of adding them based on "Two blank lines before starting static blocks, fields, constructors and inner classes
" (https://www.baeldung.com/java-clean-code#4-whitespaceshttps://www.baeldung.com/java-clean-code#4-whitespaces)

**Comment by mpellicer** [comment]
> @mpellicer Do you mind those extra lines a lot? I got in the habit of adding them based on "Two blank lines before starting static blocks, fields, constructors and inner classes" https://www.baeldung.com/java-clean-code#4-whitespaces

I don't mind much, but it's a pattern that was never adopted by Sakai developers, I'm fine leaving them as is.

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/servlet/evaluation/ExportReportServlet.java [hunk 21]
```diff
@@ -0,0 +1,353 @@
...
+    public static final String PARAM_SITE_ID = "siteId";
+    public static final String PARAM_ASSESSMENT_ID = "assessmentId";
+    public static final String PARAM_FORMAT = "format";
+    public static final String PARAM_TYPE = "type";
+    public static final String EXPORT_TYPE_ITEM_ANALYSIS = "item_analysis";
+    public static final String EXPORT_TYPE_STATISTICS = "statistics";
+
+    private static final ResourceLoader EVENT_LOG_BUNGLE = new ResourceLoader(SamigoConstants.EVAL_BUNDLE);
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/servlet/evaluation/ExportReportServlet.java:55]: Maybe 'EVENT_LOG_BUNDLE' ?

Comment (stetsche) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/servlet/evaluation/ExportReportServlet.java:55]: Might be more descriptive xD

**Comment by stetsche** [comment]
> I don't mind much, but it's a pattern that was never adopted by Sakai developers, I'm fine leaving them as is.

I see, though pretty much all files I added this year have this pattern, so I feel like I might just continue.

**Review by mpellicer — APPROVED** [review]
Looks an amazing piece of work, well done @stetsche

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/EvaluationMessages_eu.properties (hunk 9 above)
Comment (PabloSanRoman) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/EvaluationMessages_eu.properties:359]: Hi @mpellicer !

print_report=Inprimatu txostena
export_as_pdf=Esportatu PDFa
export_as_xlsx=Esportatu Excela
export_pdf=PDF gisa
export_xlsx=Excel gisa
export=Esportatu
skew_coef=Asimetria-koefizientea.
difficulty=Zailtasuna
total_attempts=Saiakerak, guztira
total_correct=Onak, guztira
total_incorrect=Txarrak, guztira

Comment (mpellicer) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/EvaluationMessages_eu.properties:359]: Thank you @PabloSanRoman !

Comment (mpellicer) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/EvaluationMessages_eu.properties:359]: @stetsche We have new translation strings

**Comment by stetsche** [comment]
Thanks @mpellicer, It should all be ready now!

**Pull request merged**